File: algorithms/0278.first_bad_version/first_bad_version.rs
// The API isBadVersion is defined for you.
// isBadVersion(versions:i32)-> bool;
// to call it use self.isBadVersion(versions)

impl Solution {
    pub fn first_bad_version(&self, n: i32) -> i32 {
		if n == 1 {
            return n;
        }
        for i in (1..n).rev() {
            if !self.isBadVersion(i) {
                return i + 1;
            }
        }
        1
    }
}

File: algorithms/0278.first_bad_version/first_bad_version.cpp
// The API isBadVersion is defined for you.
// bool isBadVersion(int version);

class Solution {
public:
    int firstBadVersion(int n) {
        int lo = 1, hi = n;
        while (lo <= hi) {
            const int mi = ((hi - lo) >> 1) + lo;
            if (isBadVersion(mi)) {
                hi = mi - 1;
            } else {
                lo = mi + 1;
            }
        }
        return lo;
    }
};